### PR TITLE
PRODUCT_UPDATED webhook

### PR DESCRIPTION
**Pull request by tomaszszymanski129**
I want to merge this change because it adds PRODUCT_UPDATED webhook.

TODO:
* [X] Add a new event to WebhookEvents
* [X] new hook to PluginManager - product_updated
* [X] new hook to BasePlugin - product_updated
* [X] new hook to WebhookPlugin - product_updated
* [X] payload should be generated by `saleor.webhook.payloads.generate_product_payload`

Where is this triggered?
* [X] `ProductUpdate` mutation
* [X] `ProductVariantCreate` mutation
* [X] `ProductVariantUpdate` mutation
* [x] `ProductSetAvailabilityForPurchase` mutation
* [x] `DeleteMetadata` mutation (selective)
* [x] `UpdatePrivateMetadata` mutation (selective)
* [x] `DeletePrivateMetadata` mutation (selective)

# Impact

* [ ] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [X] Privileged queries and mutations are guarded by proper permission checks
* [X] Database queries are optimized and the number of queries is constant
* [X] Database migration files are up to date
* [x] The changes are tested
* [X] GraphQL schema and type definitions are up to date
* [x] Changes are mentioned in the changelog

### CODE REVIEW

**Review by IKarbowiak — APPROVED** [review]

File: saleor/graphql/product/mutations/products.py [hunk 1]
```diff
@@ -1055,6 +1055,7 @@ def save(cls, info, instance, cleaned_input):
         attributes = cleaned_input.get("attributes")
         if attributes:
             AttributeAssignmentMixin.save(instance, attributes)
+        info.context.plugins.product_updated(instance)
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:1058]: Is `order_updated` called when order is created? If yes, I assume we should have the same behaviour.
What about an update in the product's variants? Do we want to also call the hook in that case? cc @maarcingebala

Comment (korycins) [reply saleor/graphql/product/mutations/products.py:1058]: What about the situation when the product has been deleted? What about the mutation for Product META?

Comment (tomaszszymanski129) [reply saleor/graphql/product/mutations/products.py:1058]: > Is `order_updated` called when order is created? If yes, I assume we should have the same behaviour.

It's not called - there is a dedicated webhook product created for that, so I figured we shouldn't call update there.

> What about an update in the product's variants? Do we want to also call the hook in that case? cc @maarcingebala 

Added that according to the ticket, it makes sense.

About deleting the product - I think there should be separate webhook for that - product deleted.

About the META - probably we should trigger the update - it's an update to a product after all.

So the plan would be...
- create a ticket for product deleted webhook
- trigger product updated when META changes

CC @maarcingebala

**Review by fowczarek — COMMENTED** [review]
Should we trigger this webhook on `ProductSetAvailabilityForPurchase` mutation?

**Comment by korycins** [comment]
> Should we trigger this webhook on `ProductSetAvailabilityForPurchase` mutation?

Yes, everything related to changes in product/its variatns

File: saleor/graphql/product/mutations/products.py [hunk 2]
```diff
@@ -1320,6 +1332,7 @@ def save(cls, info, instance, cleaned_input):
             AttributeAssignmentMixin.save(instance, attributes)
             instance.name = generate_name_for_variant(instance)
             instance.save(update_fields=["name"])
+        info.context.plugins.product_updated(instance.product)
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:1335]: It seems that there are a lot of places where we could prefetch the related objects. Maybe we could move it `PluginManager.product_updated` ? What do you think?

**Review by maarcingebala — APPROVED** [review]

File: saleor/graphql/product/mutations/products.py [hunk 3]
```diff
@@ -1890,4 +1903,5 @@ def perform_mutation(cls, _root, info, **data):
             product.available_for_purchase = start_date
 
         product.save(update_fields=["available_for_purchase", "updated_at"])
+        info.context.plugins.product_updated(product)
```
Comment (fowczarek) [inline saleor/graphql/product/mutations/products.py:1906]: We should prefetch `variants` in this mutation.

Comment (tomaszszymanski129) [reply saleor/graphql/product/mutations/products.py:1906]: yep, im adding that as we speak

**Comment by fowczarek** [comment]
@korycins @maarcingebala 
Maby `info.context.plugins.product_updated(product)` should get only product ID and fetch all needed data in an optimal way, instead of prefetching variants in all mutations witch trigger this webhook?

**Review by fowczarek — APPROVED** [review]

**Review by korycins — APPROVED** [review]

**Comment by tomaszszymanski129** [comment]
Discussed with @fowczarek and @korycins, prefetching does no performance improvement since we call variants.all() later, there is a problem with the serializer that collects data for order (quantity is no longer on variant), that will be handled in subsequent PR.

File: saleor/graphql/meta/mutations.py [hunk 4]
```diff
@@ -119,7 +119,21 @@ def perform_mutation(cls, root, info, **data):
         return cls.success_response(instance)
 
 
-class DeleteMetadata(BaseMetadataMutation):
+class ExtraMethodMetadataMixin:
```
Comment (fowczarek) [inline saleor/graphql/meta/mutations.py:122]: Should we add this mixin to `UpdateMetadata` ?

**Review by fowczarek — APPROVED** [review]

**Pull request merged**